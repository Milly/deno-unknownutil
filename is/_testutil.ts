// It seems 'IsExact' in deno_std is false positive so use `Equal` in type-challenges
// https://github.com/type-challenges/type-challenges/blob/e77262dba62e9254451f661cb4fe5517ffd1d933/utils/index.d.ts#L7-L9
export type Equal<X, Y> = (<T>() => T extends X ? 1 : 2) extends
  (<T>() => T extends Y ? 1 : 2) ? true : false;

export function stringify(x: unknown): string {
  if (x instanceof Date) return `Date(${x.valueOf()})`;
  if (x instanceof Promise) return "Promise";
  if (typeof x === "function") return x.toString();
  if (typeof x === "bigint") return `${x}n`;
  if (typeof x === "symbol") return x.toString();
  return JSON.stringify(x);
}
