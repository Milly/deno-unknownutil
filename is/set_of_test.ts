import { assertEquals } from "@std/assert";
import { assertSnapshot } from "@std/testing/snapshot";
import { assertType } from "@std/testing/types";
import { type Equal, testWithExamples } from "../_testutil.ts";
import { is } from "./mod.ts";
import { isSetOf } from "./set_of.ts";

Deno.test("isSetOf<T>", async (t) => {
  await t.step("returns properly named function", async (t) => {
    await assertSnapshot(t, isSetOf(is.Number).name);
    await assertSnapshot(t, isSetOf((_x): _x is string => false).name);
  });
  await t.step("returns proper type predicate", () => {
    const a: unknown = new Set([0, 1, 2]);
    if (isSetOf(is.Number)(a)) {
      assertType<Equal<typeof a, Set<number>>>(true);
    }
  });
  await t.step("returns true on T set", () => {
    assertEquals(isSetOf(is.Number)(new Set([0, 1, 2])), true);
    assertEquals(isSetOf(is.String)(new Set(["a", "b", "c"])), true);
    assertEquals(isSetOf(is.Boolean)(new Set([true, false, true])), true);
  });
  await t.step("returns false on non T set", () => {
    assertEquals(isSetOf(is.String)(new Set([0, 1, 2])), false);
    assertEquals(isSetOf(is.Number)(new Set(["a", "b", "c"])), false);
    assertEquals(isSetOf(is.String)(new Set([true, false, true])), false);
  });
  await testWithExamples(t, isSetOf((_: unknown): _ is unknown => true), {
    excludeExamples: ["set"],
  });
});
