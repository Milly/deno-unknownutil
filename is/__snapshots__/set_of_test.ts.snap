export const snapshot = {};

snapshot[`isSetOf<T> > returns properly named function 1`] = `"isSetOf(isNumber)"`;

snapshot[`isSetOf<T> > returns properly named function 2`] = `"isSetOf((anonymous))"`;
